ui/core/src/utils/time-series-data.ts

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { AbsoluteTimeRange, TimeScale, TimeSeries, TimeSeriesData } from '../model';
+import { gcd } from './mathjs';
+
+export const MIN_STEP_INTERVAL_MS = 10;
+
+/**
+ * Given a common time scale (see `getCommonTimeScale`), generates an array of
+ * timestamp values in ms for the x axis of a graph.
+ */
+export function getXValues(timeScale: TimeScale): number[] {
+  const xValues: number[] = [];
+  let timestamp = timeScale.startMs;
+  while (timestamp <= timeScale.endMs) {
+    xValues.push(timestamp);
+    timestamp += timeScale.stepMs;
+  }
+  return xValues;
+}
+
+/**
+ * Given a TimeSeries from a query and a common time scale (see `getCommonTimeScale`),
+ * gets the values for the y axis of a graph, filling in any timestamps that are
+ * missing from the time series data with `null` values.
+ */
+export function getYValues(series: TimeSeries, timeScale: TimeScale): Array<number | null> {
+  let timestamp = timeScale.startMs;
+
+  const yValues: Array<number | null> = [];
+  for (const valueTuple of series.values) {
+    // Fill in values up to the current series value timestamp with nulls
+    while (timestamp < valueTuple[0]) {
+      yValues.push(null);
+      timestamp += timeScale.stepMs;
+    }
+
+    // Now add the current value since timestamp should match
+    yValues.push(valueTuple[1]);
+    timestamp += timeScale.stepMs;
+  }
+
+  // Add null values at the end of the series if necessary
+  while (timestamp <= timeScale.endMs) {
+    yValues.push(null);
+    timestamp += timeScale.stepMs;
+  }
+
+  return yValues;
+}
+
+/**
+ * Given a list of running queries, calculates a common time scale for use on
+ * the x axis (i.e. start/end dates and a step that is divisible into all of
+ * the queries' steps).
+ */
+export function getCommonTimeScale(seriesData: Array<TimeSeriesData | undefined>): TimeScale | undefined {
```

Comment:
I modified this function to take series data to make it more generic and unopinionated about where the data came from, since this is being exported for more generic use cases. That also avoids having to teach `@perses-dev/core` about react-query. I left a small method that wraps this in `data-transform` that still handles the query part. Let me know if you think there's a better abstraction here.